src/main/scala/org/broadinstitute/hail/driver/FilterAlleles.scala

```diff
@@ -0,0 +1,158 @@
+
+import scala.collection.{GenTraversableOnce, mutable}
+import scala.reflect.ClassTag
+
+object FilterAlleles extends Command {
+  class Options extends BaseOptions {
+    @Args4jOption(required = false, name = "--keep", usage = "Keep variants matching condition")
+    var keep: Boolean = false
+
+    @Args4jOption(required = false, name = "--remove", usage = "Remove variants matching condition")
+    var remove: Boolean = false
+
+    @Args4jOption(required = true, name = "-c", aliases = Array("--condition"),
+      usage = "Filter expression involving <code>v</code> (variant), <code>va</code> (variant annotations), and <code>aIndex</code> (allele index)",
+      metaVar = "COND_EXPR")
+    var condition: String = _
+
+    @Args4jOption(required = true, name = "-a", aliases = Array("--annotation"),
+      usage = "Annotation modifying expression involving <code>v</code> (new variant), <code>va</code> (old variant annotations), and <code>aIndices</code> (maps from new to old indices)",
+      metaVar = "ANNO_EXPR")
+    var annotation: String = _
+  }
+
+  override def newOptions: Options = new Options
+
+  override def name: String = "filteralleles"
+
+  override def description: String = "Filter alleles in current dataset using the Hail expression language"
+
+  override def supportsMultiallelic: Boolean = true
+
+  override def requiresVDS: Boolean = true
+
+  override protected def run(state: State, options: Options): State = {
+    if ((options.keep && options.remove) || (!options.keep && !options.remove))
+      fatal("one `--keep' or `--remove' required, but not both")
+
+    val conditionEC = EvalContext(Map(
+      "v" -> (0, TVariant),
+      "va" -> (1, state.vds.vaSignature),
+      "aIndex" -> (2, TInt)))
+    val conditionE = Parser.parse[Boolean](options.condition, conditionEC, TBoolean)
+
+    val annotationEC = EvalContext(Map(
+      "v" -> (0, TVariant),
+      "va" -> (1, state.vds.vaSignature),
+      "aIndices" -> (2, TArray(TInt))))
+    val (types, generators) = Parser.parseAnnotationArgs(options.annotation, annotationEC, Annotation.VARIANT_HEAD)
+    val inserterBuilder = mutable.ArrayBuilder.make[Inserter]
+    val finalType = types.foldLeft(state.vds.vaSignature) { case (vas, (path, signature)) =>
+      val (newVas, i) = vas.insert(signature, path)
+      inserterBuilder += i
+      newVas
+    }
+    val inserters = inserterBuilder.result()
+
+    val keep = options.keep
+
+    def filterAllelesInVariant(v: Variant, va: Annotation): Option[(Variant, IndexedSeq[Int], Array[Int])] = {
+      var allZero = true
+      var dead = 0
+      val oldToNew =
+        coalesce(v.altAlleles)(v.altAlleles.size+1, (_, altIndex) => altIndex + 1, 0) { (_,_,_,index) =>
+            conditionEC.setAll(v, va, index)
+            if (Filter.keepThis(conditionE(), keep)) {
+              allZero = false
+              index - dead
+            } else {
+              dead += 1
+              0
+            } }
```

Comment:
I think this may be unnecessarily complicated.  

`val passAlleles: Array[(AltAllele, Int)] = v.AltAlleles.zipWithIndex.filter(...)` will give you all this information pretty easily, I think.  Then:

`if (allZero)` becomes `if (passAlleles.isEmpty)`

```scala
val oldtoNew: Array[Option[Int]] = {
  val m = passAlleles.zipWithIndex
    .map { case ((_, oldIndex), newIndex) => (newIndex, oldIndex) }
    .toMap
  Array.tablulate(v.nAltAlleles)(i => m.get(i)) 
}
```
etc